network/src/connection_manager.rs

```diff
@@ -0,0 +1,404 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{
+    external::{
+        message_types::{GetPeers, Version},
+        Channel,
+        PingPongManager,
+    },
+    internal::{
+        context::{Connections, Context},
+        PeerBook,
+        PeerInfo,
+    },
+    RequestManager,
+};
+use snarkos_consensus::MerkleTreeLedger;
+
+use chrono::Utc;
+use std::{collections::HashMap, net::SocketAddr, sync::Arc};
+use tokio::sync::RwLock;
+
+#[derive(Clone)]
```

Comment:
it might be undesirable for `ConnectionManager` to derive `Clone`, as it contains a `HashMap` and `Vec` that will get deep-cloned, incurring extra allocations; cloned instances could also become out-of-sync wrt. those two fields; it might be desirable, just making sure